polytools.degree: Speed up by avoiding Poly term collection

user234683 opened the pull request:
`Poly(f, var)` can sometimes take much longer than `Poly(f)` because it collects with respect to the variable. This behavior is unnecessary for `degree`, so this commit uses `Poly(f)` instead.

Before:
```
>>> import timeit
>>> timeit.timeit("degree(f, c5)", globals=globals(),number=20)
77.93264110297285
>>> timeit.timeit("Poly(f).degree(c5)", globals=globals(),number=20)
2.8491278389856802
```
After:
```
>>> timeit.timeit("degree(f, c5)", globals=globals(),number=20)
2.906191912352824
>>> timeit.timeit("Poly(f).degree(c5)", globals=globals(),number=20)
2.888086826285871
```

<!-- Please give this pull request a descriptive title. Pull requests with descriptive titles are more likely to receive reviews. Describe what you changed! A title that only references an issue number is not descriptive. -->

<!-- If this pull request fixes an issue please indicate which issue by typing "Fixes #NNNN" below. -->
See sympy/sympy#12998. This fixes it for `degree`, but not `degree_list`

Commit:
degree: Speed up by avoiding Poly term collection

`Poly(f, var)` can sometimes take much longer than `Poly(f)` because it collects with respect to the variable. This behavior is unnecessary for `degree`, so this commit uses `Poly(f)` instead. See sympy/sympy#12998

Before:
```
>>> import timeit
>>> timeit.timeit("degree(f, c5)", globals=globals(),number=20)
77.93264110297285
>>> timeit.timeit("Poly(f).degree(c5)", globals=globals(),number=20)
2.8491278389856802
```
After:
```
>>> timeit.timeit("degree(f, c5)", globals=globals(),number=20)
2.906191912352824
>>> timeit.timeit("Poly(f).degree(c5)", globals=globals(),number=20)
2.888086826285871
```

asmeurer commented:
Some test failures in Travis.

Commit:
Fixed error when multiple gens are given

Commit:
degree: test that gens not in expression give 0

An error almost slipped in where degree(x*y**2, z) rose an exception, rather than giving 0. Add a test to make sure that won't be able to happen.
Also added test for degree(0,x) == -oo for good measure.

jksuom commented on sympy/polys/polytools.py:
```diff
@@ -4404,13 +4404,20 @@ def degree(f, *gens, **args):
     """
     options.allowed_flags(args, ['gen', 'polys'])
-
+    f = sympify(f).as_expr()
```
Is `as_expr` necessary? (`poly_from_expr` will accept `Poly`s as well.)

user234683 replied:
If `f` is a poly, it might not have the gen for the variable whose degree we need (and poly_from_expr will retain the gens that the input has). For instance, suppose we call `degree(Poly(x**2+y**2, x), y)`. `poly_from_expr(Poly(x**2+y**2,x))[0]` gives back `Poly(x**2+y**2,x)` (same gens). If we tried to then call `Poly(x**2+y**2,x).degree(y)`, it would give an error because y isn't one of the gens.

jksuom replied:
Ok, thanks.

jksuom commented on sympy/polys/polytools.py:
```diff
@@ -4404,13 +4404,20 @@ def degree(f, *gens, **args):
+    f = sympify(f).as_expr()
     try:
-        F, opt = poly_from_expr(f, *gens, **args)
+        if f.is_Number:
```
Why should this case be treated differently?

user234683 replied:
Poly() won't take a number without generators

jksuom replied:
Could we just return `S.Zero` or `S.NegativeInfinity`?

user234683 replied:
Yeah, but I think speed improvement would be negligible (I just passed it to `Poly` with gens when it's a number so I didn't have to re-code the logic, since `Poly` won't spend time collecting any terms for a number.) But I'll make another commit if it makes the code clearer

Commit:
polytools.degree: fixed numerous corner cases

See the tests added to polys/tests/test_polytools.py

Commit:
polytools: added tests for corner cases for degree

Commit:
Fixed minor ComputationFailed string bug

Was doing this:
`sympy.polys.polyerrors.ComputationFailed: <unprintable ComputationFailed object>`

Commit:
Fixed another corner case

`degree(Poly(0,x),z)` would give `sympy.polys.polyerrors.ComputationFailed: degree(0) failed without generators`
Also made the code slightly less like spaghetti (by appending gens from Polys instead of doing special actions in response to checking if given gens are in those Polys gens).

Commit:
polytools: degree: test for two more corner cases

user234683 commented:
The code is a bit more messy because of the optional `gen` keyword arg. In my opinion,`degree` should only have only one optional argument, like so: `def degree(f, gen=None)` or `def degree(f, *gens)`.

Also, while I've maintained all cases where the old degree works (plus errors that have tests), there's some times where the old degree would error and this version does not:
```
>>> degree(0, gen=x)
-oo
>>> old_degree(0, gen=x)
degree(0) failed without generators

>>> degree(y**2 + x**3, x, gen=y)
2
>>> old_degree(y**2 + x**3, x, gen=y)
a valid generator expected, got y

>>> degree(Poly(y**2 + x**3, y), x, gen=y)
2
>>> old_degree(Poly(y**2 + x**3, y), x, gen=y)
a valid generator expected, got y

>>> degree(Poly(y**2+x**3,x), gen=y)
2
>>> old_degree(Poly(y**2+x**3,x), gen=y)
a valid generator expected, got y
```
Replicating all these cases would make the code significantly more messy and hard to follow. This is another reason why I advocate deprecating the `gen` keyword argument.

jksuom commented on sympy/polys/polytools.py:
```diff
@@ -4404,13 +4404,34 @@ def degree(f, *gens, **args):
     options.allowed_flags(args, ['gen', 'polys'])
+    f, orig = sympify(f), f
+    gens = list(gens)
+    try:
```
Is is possible to use something like this instead of a try block:
```
gen = args.get('gen')
if gen:
    gens = [sympify(gen)] + gens

user234683 replied:
Done

Commit:
polys.degree: if statement instead of try

user234683 commented:
I found something that differs from the previous implementation:
```
>>> old_degree(Poly(x, x),gen=0)
1
>>> degree(Poly(x, x),gen=0)
0
>>> old_degree(x,gen=0)
1
>>> degree(x,gen=0)
0
```
In old_degree, gen is interpreted as the index of a gen instead of the gen if it's a number. Should I make this work? It has tests for it but with Poly.degree. It would probably make the code pretty convoluted. Another reason I'd suggest deprecating the gen keyword.

jksuom commented:
It seems that `degree` is intended to return the degree in a single variable and that `gen` is the principal way of defining that variable. How would `degree` work if there were no `gen`?

user234683 commented:
At the moment the function signature is `def degree(f, *gens, **args):`
It accepts any number of gens (only the first is actually used), and additionally, has the `gen` keyword argument in `args` (which will override `gens` if given unless `gen` is not in `gens`, in which case the old implementation raises an error.)
So all of these are equivalent:
```
degree(x**2 + y**3, x)
degree(x**2 + y**3, gen=x)
degree(x**2 + y**3, x, gen=x)
degree(x**2 + y**3, x, y, gen=x)
degree(x**2 + y**3, y, x, gen=x)
old_degree(x**2 + y**3, x, y, gen=0)
old_degree(x**2 + y**3, y, x, gen=1)
```
My view is that we shouldn't have both `gens` and `gen`, and that one or the other should be deprecated (I side toward deprecating `gen`).

Commit:
polys.degree: Fixed newlines (hopefully)

Mixed Unix and Windows newline conventions were causing a syntax error

smichr commented:
> only the first is actually used

All gens are used in the call `poly_from_expr(f, *gens, **args)` and then the generator to use when calculating the degree is specified by gen (which in the Poly.degree method defaults to the first generator).

jksuom commented:
I agree that having both `gens` and `gen` looks confusing but I'd hesitate to remove one or the other. In particular, I think that it should be possible to write `poly.degree(i)`, where `i` is either an index or a generator, in the form `degree(poly, gen=i)`.

Perhaps the following strategy could be applied. If `gen` is given, ignore `gens`. If there is no `gen`, take the first item of `gens` as `gen` and ignore the rest. If neither `gen` nor `gens` is given, set `gen=0` (the default).

smichr commented:
As much as possible, function `degree` should not duplicate what method `degree` will already do. I suggest the following:

```
def degree(f, *gens, **args):
    """
    Return the degree of ``f`` in the given variable.

    The degree of 0 is negative infinity.

    Examples
    ========

    >>> from sympy import degree
    >>> from sympy.abc import x, y

    >>> degree(x**2 + y*x + 1, gen=x)
    2
    >>> degree(x**2 + y*x + 1, gen=y)
    1
    >>> degree(0, x)
    -oo

    """
    options.allowed_flags(args, ['gen', 'polys'])
    gen = args.pop('gen', 0)
    symbolic_gen = isinstance(gen, Expr) and gen.free_symbols
    if not isinstance(f, Poly):
        f = sympify(f)
        if f.is_Number:
          if not gens and not symbolic_gen:
            raise ComputationFailed('generators or a gen with free symbols must be give')
          f = Poly(f, gens or gen)
    if gens or not isinstance(f, Poly):
      try:
        if not f.is_Number:
          F, opt = poly_from_expr(f, **args)
        else:
          F, opt = poly_from_expr(f, gens, **args)
      except PolificationFailed as exc:
        raise ComputationFailed('degree', 1, exc)
    else:
      F = f
    return F.degree(gen, **args)
```

The test
```
assert degree(Poly(exp(x) + y**2, y), exp(x)) == 1
```
could be added. This is a case where an explicit gens is given and is needed to override the orginal Poly definition; gen defaults to 0 and with len(gens) == 1, this must correspond to exp(x).

smichr commented:
> I'd hesitate to remove one or the other

agree -- gens is there to make an expr into a Poly; gen is there to indicate which generator to use.

smichr commented:
Here are some untested tests for the code snippet I added:

```
assert degree(0, gen=x) == Poly(S(0), x).degree()
assert degree(0, x) == Poly(S(0), x).degree()
assert degree(1, gen=x) == Poly(S(1), x).degree()
assert degree(1, x) == Poly(S(1), x).degree()
assert degree(Poly(exp(x) + y**2, y), exp(x)) == 1
assert degree(Poly(exp(x) + y**2, y), y, exp(x), gen=1) == 1
p = exp(x) + y**2
poly = Poly(p)
assert degree(p) == Poly(p).degree(poly.gens[0])
assert degree(p, 1) == Poly(p).degree(poly.gens[1])
# needs gens or symbolic gen
raises(ComputationFailed, lambda: degree(0))
# designated gen is not amongst the given gens
raises(ComputationFailed, lambda: Poly(exp(x) + y**2, y), exp(x), gen=y))
raises(ComputationFailed, lambda: degree(1, x, gen=y))
```

Commit:
polytools.degree: gen kwarg can now take an index

In addition to symbols/expression to specify the gen, gen can now take an integer index for expressions which already have gens or for which they can be determined.

user234683 commented:
>As much as possible, function degree should not duplicate what method degree will already do. I suggest the following:

I tested it out but there's some bugs:
```
>>> degree(x*y**2, y, x)
1
>>> old_degree(x*y**2, y, x)
2
>>> degree(x*y**2, z)
1
>>> old_degree(x*y**2, z)
0
>>> degree(Poly(x**2+y, y), x)
1
>>> old_degree(Poly(x**2+y, y), x)
2
```
>agree -- gens is there to make an expr into a Poly; gen is there to indicate which generator to use.

Actually, gens cannot be used anymore to make the expr into a Poly (since that is what causes the slow speed). We have to call Poly(f) without gens. To preserve backwards compatibility, the logic of how these arguments interact has to be recreated.

Making the expr into a Poly to me is an implementation detail. Suppose we didn't have a `Poly` class. There would certainly still be a degree function, and it would presumably crawl through the expression to determine the degree. Using a custom method to find the degree isn't hypothetical. Consider something like this: https://github.com/sympy/sympy/pull/7236
Having these additional ways to specify the variable which mesh together in complex ways when both ways are used makes implementing things like this unnecessarily complex.

From the user's point of view, it doesn't matter if degree needs to make a Poly or if it turns out faster to use some specialized method to get the degree. The user is only interested in getting the degree of a specific variable.

Having both `*gens` and `gen=` essentially act as a feature to let the user give a list of 'allowed' variables that may or may not have what they're interested in, and then provide the one they're actually interested in using the `gen=` argument, but raise an error if that one isn't in their list of 'allowed' ones. It's a very strange feature. If for some reason the user needed to place a restriction on themself about which variables are allowed, they could write that logic where it's needed.

Of course, I'm not proposing we immediately remove one or the other (because then code would break). My opinion is that it should raise a deprecation warning and be removed in the future.

I committed an implementation that should now work with integer `gen=` arguments. It so happens that it also now preserves some cases where the canonical implementation raises errors:
```
>>> degree(0, gen=x)
degree(0) failed without generators
>>> old_degree(0, gen=x)
degree(0) failed without generators

>>> degree(Poly(y**2+x**3,x), gen=y)
a valid generator expected, got y
>>> old_degree(Poly(y**2+x**3,x), gen=y)
a valid generator expected, got y
```
Though it still succeeds in these cases whereas the canonical one does not:
```
>>> degree(y**2 + x**3, x, gen=y)
2
>>> old_degree(y**2 + x**3, x, gen=y)
a valid generator expected, got y
>>> degree(Poly(y**2 + x**3, y), x, gen=y)
2
>>> old_degree(Poly(y**2 + x**3, y), x, gen=y)
a valid generator expected, got y
```
These could also be made to raise errors with an additional `if` statement if it's important.

What do ya'll think about raising a deprecation warning?

>Here are some untested tests for the code snippet I added:
```
assert degree(0, gen=x) == Poly(S(0), x).degree()
assert degree(1, gen=x) == Poly(S(1), x).degree()
```

This actually doesn't hold for the canonical implementation, so it's not strictly necessary for keeping the same behavior (unless there's consensus to change it.)

```
raises(ComputationFailed, lambda: Poly(exp(x) + y**2, y), exp(x), gen=y))
raises(ComputationFailed, lambda: degree(1, x, gen=y))
```

Current implementation actually raises `PolynomialError`. I'll add the other tests

Commit:
degree: Added more tests

smichr commented:
To make the method and function behave similarly, I think it would make more sense to have the function signature be `degree(f, gen=0, **kwargs)` and to retrieve `gens` from `kwargs`. gens are not needed when calling the method because they are already part of the Poly, `self`. It's awkward to put `*gens` in the signature for the function because (as you are pointing out) these are not needed and, in fact, shouldn't be used since the gen itself is sufficient to both create the Poly and find the degree. The only exception is when the `gen` is given as an integer. Then the gens have to be given--or `f` should already be a Poly and the user should really be calling `f.degree(gen=some_integer)`.

smichr commented:
(Sorry, that last post was made before reading your last post. I'll look at this some more tomorrow.)

user234683 commented:
>To make the method and function behave similarly, I think it would make more sense to have the function signature be degree(f, gen=0, **kwargs) and to retrieve gens from kwargs. gens are not needed when calling the method because they are already part of the Poly, self. It's awkward to put *gens in the signature for the function because (as you are pointing out) these are not needed and, in fact, shouldn't be used since the gen itself is sufficient to both create the Poly and find the degree. The only exception is when the gen is given as an integer. Then the gens have to be given--or f should already be a Poly and the user should really be calling f.degree(gen=some_integer).

I agree with this. Though I'd go so far to say gens wouldn't even be needed in `**kwargs`. If the user wants to select a gen from a list based on its index, it's not necessary to have them pass the list and index in and do it for them. Since they have the list, and the index, they can simply use `gens[index]` themself and pass that in instead. If `f` is a `Poly`, and gen is an integer, it would then, as you said, select from that poly's gens by index. If it's not a `Poly`, it would presumably just go based on alphabetic order as it does now.
I guess what needs to be figured out is how to transition to such a function signature.

smichr commented:
Here's another draft. The purpose of gens would be as a double check -- if gen isn't there, an error would be raised. So in this sense it provides a safety similar to Poly.degree:

```python
def degree(f, gen=0, **kw):
  if isinstance(f, Poly):
    f = f.as_expr()  # it gets messier to leave this as a Poly with its preexisting gens
  else:
    f = sympify(f).expand() # mul and multinomial
  if f.is_number:
    assert f.is_zero is not None
    return S.NegativeInfinity if f.is_zero else S.Zero
  f = Poly(f) # or use poly_from_expr
  gens = kw.get('gens', None)
  try:
    gen = as_int(gen)
  except ValueError:
    if gens and gen not in gens:
      gen = len(gens)  # raise unified error below in assertion
    else:                                      # |
      gen = gens.index(gen)                    # |
  if gens:  # this is a redundant, hence strict# |
    assert -len(gens) <= gen < len(gens)       #<+
    assert gens[gen] in f.gens # be cautious if they gave gens
    gen = f.gens.index(gens[gen])
  try:
    return f.degree(gen)
  except PolynomialError:
    return S.Zero  # generator not present or index out of range
```

smichr commented:
FYI: my code posts don't imply that they are better or should be used. It is my way of thinking through the issues.



I think that part of the issue here is whether degree should behave like a thin wrapper to Poly.degree. Poly degree doesn't need generators because they are already identified; it is strict in that if a generator is not present it complains. So if the function is to behave the same way, all you have to do is return `Poly(f.as_expr().expand()).degree(args.get('gen', 0))` except for when f is a number. If, on the other hand, Poly.degree is just a method of computing the degree of a polynomial then it might make sense to make it less strict: if the requested gen does not exist then 0 could be returned (or -oo in the case of the expression being 0). So maybe this question should be settled first: what will be done with a gen which is either out of bounds or doesn't exist?

smichr commented on sympy/polys/polytools.py:
```diff
@@ -4404,13 +4404,33 @@ def degree(f, *gens, **args):
+                                  (len(gens), len(gens), gen))
+    if gen in F.gens:
+        return sympify(F.degree(gen))
+    return S.Zero
```
I'm not sure it's a good idea that degree and Poly.degree return different types. I think there are other functions that return ints. I see that you aren't changing anything, but perhaps this is an issue to look into.

user234683 replied:
Yeah I was wondering why sympify was there in the current implentation.

user234683 commented:
>So maybe this question should be settled first: what will be done with a gen which is either out of bounds or doesn't exist?

In my view, numbered gens should raise an error if they're out of bounds. Better to fail explicitly instead of silently. Symbolic gens should always work, regardless of whether the input is given as a poly and doesn't contain it.

Here's an implementation that has a single gen argument, which defaults to 0:
```python
@public
def degree(f, gen=0, **args):
    """
    Return the degree of ``f`` in the given variable.

    The degree of 0 is negative infinity.

    Examples
    ========

    >>> from sympy import degree
    >>> from sympy.abc import x, y

    >>> degree(x**2 + y*x + 1, gen=x)
    2
    >>> degree(x**2 + y*x + 1, gen=y)
    1
    >>> degree(0, x)
    -oo

    """
    options.allowed_flags(args, ['polys'])
    f = sympify(f)
    if isinstance(gen, int):
        if f.is_Poly:
            return sympify(f.degree(gen))
        try:
            F, opt = poly_from_expr(f, **args)
        except PolificationFailed as exc:
            raise ComputationFailed('degree', 1, exc)
        return sympify(F.degree(gen))
    
    f = f.as_expr()
    try:
        if f.is_Number:
           F, opt = poly_from_expr(f, gen, **args)
        else:
           F, opt = poly_from_expr(f, **args)
    except PolificationFailed as exc:
        raise ComputationFailed('degree', 1, exc)

    if gen in F.gens:
        return sympify(F.degree(gen))
    return S.Zero
```
I looked through some files that import the degree function. (I got them using `(import \([^\)]*,[ \n\r]*degree|import degree|import [^\n\r]*, degree|import \(degree)` ), and so far they all seem to use `degree(poly, var)`. I haven't seen any using indices or even using the gen kwarg. But I'm going to try running the tests to be sure.

Also, does anyone know what the `polys` kwarg does?

smichr commented:
> anyone know what the polys kwarg does

When appropriate, it returns the result as a Poly instead of an expr. I don't think it has any meaning in this context as I can't imagine anything but an integer coming back:

```
>>> minpoly(sqrt(y),x)
x**2 - y
>>> minpoly(sqrt(y),x,polys=True)
Poly(x**2 - y, x, domain='ZZ(y)')
```

smichr commented:
> Here's an implementation that has a single gen argument, which defaults to 0:

I think that looks really good. I would drop `args` as it is not needed (nor does it need to be passed along to `poly_from_expr`. On the otherhand, if someone knows that the expression shouldn't be expanded, the `expand=False` flag could be sent -- though that is not recognized as a valid option right now. Since it's not recognized now, I don't see the harm in dropping it.

2 comments:

- why `Number` instead of `number`?
- you have to handle the `is_zero` issue or else you will silently return an invalid result

smichr commented:
What about this:

```
def degree(f, gen=0):
  if isinstance(gen, int) and isinstance(f, Poly):
    F = f
  else:
    F = sympify(f).as_expr()
    try:
        if F.is_Number:
            x = Dummy() if isinstance(gen, int) else gen
            F, opt = poly_from_expr(F, x)
        else:
            F, opt = poly_from_expr(F)
            if not (isinstance(gen, int) or gen in F.gens):
              F = poly_from_expr(Dummy(), gen)
    except PolificationFailed as exc:
        raise ComputationFailed('degree', 1, exc)

  return F.degree(gen)
```

smichr commented:
Do you think you could update the docstring of degree to have a see also for `total_degree`, the need for which is given in #12007?

smichr commented:
cf #6322, though this is only tangentially related

user234683 commented:
>What about this:

It's a bit hard to follow. But works otherwise. For instance, this is an odd way to do it:
`F, opt = poly_from_expr(Dummy(), gen)`

>Do you think you could update the docstring of degree to have a see also for total_degree, the need for which is given in #12007?

Not sure what you mean here. A docstring to demonstrate that `degree` is different than `total_degree`?

smichr commented:
> this is an odd way to do it:

The gen is symbolic but not in f.gens so we create the simplest non-numeric poly having gen in gens.

The goal is to keep degree method and function aligned wrt output.

smichr commented:
> Not sure what you mean here

Just a

```
See also
========
total_degree
```

smichr commented:
Does this close #7069, too?

user234683 commented:
I've modified it so it closes #7069. This actually meant that the try-except block could be removed since there's no scenario where it could raise an error anymore. I've also removed the `**args` and added the "see also":
```python
@public
def degree(f, gen=0):
    """
    Return the degree of ``f`` in the given variable.

    The degree of 0 is negative infinity.

    Examples
    ========

    >>> from sympy import degree
    >>> from sympy.abc import x, y

    >>> degree(x**2 + y*x + 1, gen=x)
    2
    >>> degree(x**2 + y*x + 1, gen=y)
    1
    >>> degree(0, x)
    -oo

    See also
    ========
    total_degree
    degree_list
    """

    f = sympify(f)
    if f.as_expr().is_Number:
        if f == 0:
            return S.NegativeInfinity
        return S.Zero

    if isinstance(gen, int):
        if f.is_Poly:
            return sympify(f.degree(gen))
        F, opt = poly_from_expr(f)
        return sympify(F.degree(gen))

    f = f.as_expr()
    F, opt = poly_from_expr(f)

    if gen in F.gens:
        return sympify(F.degree(gen))
    return S.Zero
```
I went to see if `degree_list` could be fixed too, and I realized there's a problem with degree in general:
```
>>> old_degree(x**4+x**2+1, x**2)
0
>>> degree(x**4+x**2+1, x**2)
0
>>> degree(exp(x)**4+exp(x)**2+1, exp(x)**2)
0
>>> degree(exp(x)**4+exp(x)**2+1, exp(x))
4
>>> degree(exp(x)**4+exp(x)**2+1, exp(x)+1)
0
>>> old_degree(exp(x)**4+exp(x)**2+1, exp(x)+1)
0
>>> old_degree(x**4+x**2+1, x+1)
0
```
Finding the degree with respect to an expression in general doesn't work (and never has). So we have an inconsistency where some gens given as expressions happen to work and others don't. I'm not certain what should be done here. Perhaps if gen is an expression it should return an error?
This issue seems to also affect `Poly`:
```
>>> Poly(x**4+x**2+1, x+1)
Poly(x**4 + x**2 + 1, x + 1, domain='ZZ[x]')
>>> Poly(x**4+x**2+1, x**2)
Poly(x**4 + x**2 + 1, x**2, domain='ZZ[x]')
>>> Poly(x**4+x**2+1, x**2).degree()
0
```

user234683 commented:
Should I commit the new interface and we merge it? Then we can open a new issue regarding the expression generators.

smichr commented:
I will look it over tonight or tomorrow.

smichr commented:
> I will look it over tonight or tomorrow

OK. As I mentioned above, I don't like the idea of duplicating the behavior of the Poly.degree method. I would like to see any decisions about return values left there and let the degree function just get the polynomial in the right form. So -- again, inspired by your own changes -- I prefer:

```python
def degree(f, gen=0):
  if isinstance(gen, int) and isinstance(f, Poly):
    F = f
  else:
    F = sympify(f).as_expr()
    if not isinstance(gen, int) and not isinstance(gen, Symbol):
        d = Dummy()
        F = F.subs(gen, d)
        if gen.free_symbols & F.free_symbols:
            raise ComputationFailed  # symbols of gen are in other generators
        gen = d
    if F.is_Number:
        # a symbolic generator is needed
        x = Dummy() if isinstance(gen, int) else gen
        F, opt = poly_from_expr(F, x)
    else:
        F, opt = poly_from_expr(F)
        # if gen is not in the generators make an equivalent
        # F which doesn't have gen since degree(y, x) == degree(x, y)
        if not (isinstance(gen, int) or gen in F.gens):
            F, opt = poly_from_expr(Dummy(), gen)
  return F.degree(gen)
```

This may handle the issue of non-symbolic generators, too.

```
degree((x+y)**2+x+y,x+y) == 2
degree((x+y)**2+x,x+y) raises error
degree(sqrt(x)+x,sqrt(x)) raises error
degree(sqrt(x)+1,sqrt(x)) == 1
```

BUT, I think handling non-symbol gen is a very bad idea because SymPy has to guess at what the user wants: should sqrt(x) + x, gen=sqrt(x) be gen + gen**2 or gen + x? Should (x+y)**2 + x, gen=x+y be gen**2 + x or gen**2 + gen - y. In face of ambiguity, don't guess. So better this:

```python
def degree(f, gen=0):
  if isinstance(gen, int) and isinstance(f, Poly):
    F = f
  else:
    F = sympify(f).as_expr()
    if not isinstance(gen, int) and not isinstance(gen, Symbol):
        raise ComputationFailed  # gen must be int or symbol
    if F.is_Number:
        # a symbolic generator is needed
        x = Dummy() if isinstance(gen, int) else gen
        F, opt = poly_from_expr(F, x)
    else:
        F, opt = poly_from_expr(F)
        # if gen is not in the generators make an equivalent
        # F which doesn't have gen since degree(y, x) == degree(x, y)
        if not (isinstance(gen, int) or gen in F.gens):
            F, opt = poly_from_expr(Dummy(), gen)
  return F.degree(gen)
```

user234683 commented:
>OK. As I mentioned above, I don't like the idea of duplicating the behavior of the Poly.degree method. I would like to see any decisions about return values left there and let the degree function just get the polynomial in the right form

It just seems too hacky for me (i.e. making Dummy expressions to coerce `Poly.degree()` to return the value we want it to). Also, I think this function is ultimately going to have to make decisions anyway about the return values that `Poly.degree()` can't be made to do. An example I mentioned a while ago was #7236, which optimizes for when the expression is given as `(x+a)**n` where `n` might be a huge number, so it doesn't have to be expanded. I personally believe this will ultimately be implemented in the future (or at least that it should). In that case, decisions about the return value will have to be made anyway.

Even further, in the future `degree` might be changed to work on things other than just polynomials. See https://en.wikipedia.org/wiki/Degree_of_a_polynomial#Computed_from_the_function_values This is simply outside the domain of `Poly.degree()`. Also, I'm not sure what kind of decisions about the return values you have in mind. Do you mean whether it returns an `int` or `Number`? Or whether `degree(0)` gives `-oo`? I think the inconsistency of returning `int`s vs `Number`s is its own issue (and affects more than the degree functions as you mentioned.) As for `degree(0)` I think the decision has been made for that. But if it does change in the future, again `degree()` presumably would have to worry about this kind of thing anyway since it might have to deal with things other than polynomials by then.

But if you believe delegating to `Poly.degree()` is a better method, we can commit that for now and it will be worked out if/when `degree` is extended to do things `Poly.degree()` can't. What do you think?

smichr commented:
> What do you think?

OK. It sounds like you see Poly's ability to give a degree is different than the general request. And I can see the validity of that logic.

I would suggest the following (untested) revision of what you proposed above:

```python
def degree(f, gen=0):
  if not isinstance(f, Poly):
    F = sympify(f)
    if F.is_Number:
      return S.Zero if F else S.NegativeInfinity
    F, _ = poly_from_expr(F)

  if not isinstance(gen, int):
    if gen not in F.gens and isinstance(f, Poly):
      # try recast without explicit gens
      F, _ = poly_from_expr(F.as_expr())
    if gen not in F.gens:
      return S.Zero

  return sympify(F.degree(gen))
```

Commit:
polytools.degree: removed **args

Removed *gens, **args, and replaced with a single argument: gen, which can be a symbol or the index of one (for expressions this is by alphabetical order, for Polys this is based on the order of its gens).

Commit:
polytools: Edited tests for new degree interface

polytools.degree now only takes a single argument. Edits the tests for it accordingly.

user234683 commented:
> I would suggest the following (untested) revision of what you proposed above

Looks good. Fixed some minor issues and committed. Also, since I think you picked up this habit from me, `is_number` should be used instead of `is_Number`. The latter returns whether it's of the Number class, so constants like `E` would be reported as not being numbers. I recently learned the difference: #13067

smichr commented:
Regarding the use of `is_Number`. I originally thought it should be the lower case version but realized that any non-rational number will have generators recognized by Poly so no generator needs to be explicitly given. Rather than including this slightly fragile logic you *could* just let the error from Poly tell you that you need to use a generator.

A test that could be added is `assert degree(pi) == 1` which is the number-equivalent of `assert degree(x) == 1`.

Commit:
test_polytools: added test for degree of constants

Commit:
polytools.degree: fixed for constant Poly's

such as degree(Poly(0,x), x)

smichr commented:
I think the current code will fail for `degree(Poly(0, x), z)`. In addition, I think it's not a good idea to let a multivariate expression get an index to a generator. Can you see any ways to improve the following?

```python
def degree(f, gen=0):
  F = f
  if not isinstance(F, Poly):
    F = sympify(F)
    if F.is_Number:
      return S.Zero if F else S.NegativeInfinity
    F, _ = poly_from_expr(F)
  if not isinstance(gen, int):
    gen = sympify(gen)
    if gen.is_number:
      try:
        gen = as_int(gen)
      except ValueError:
        raise TypeError('a numeric generator must be an integer')
  if not isinstance(gen, int):
    if gen not in F.gens and isinstance(f, Poly):
      F = F.as_expr()
      if F.is_Number:
        return S.Zero if F else S.NegativeInfinity
      # try recast without explicit gens
      F, _ = poly_from_expr(F)
      if gen not in F.gens:
        return S.Zero
  elif not isinstance(f, Poly) and len(F.gens) > 1:
    raise ValueError(filldedent('''
For multivariate expressions, an explicit generator
must be given for gen, not an integer.
    '''))
  return sympify(F.degree(gen))
  
raises(ValueError, lambda: degree(x + y**2))
raises(ValueError, lambda: degree(pi + E**2))
raises(TypeError, lambda: degree(x, 1.1))
```

Commit:
polytools.degree: Fixed for constant Polys. Disallow non-int numeric gen and numeric gen on multivarate expr

Fixed `degree(Poly(0, x), z)` failing with `sympy.polys.polyerrors.PolificationFailed: can't construct a polynomial from 0`
Calling `degree` with a numeric but non-integer gen is now an error: `degree(x, 1.2)` -> `sympy.polys.polyerrors.PolynomialError: a valid generator expected, got 1.2`
Calling `degree` with a multivariate expression and with an integer for `gen` is now an error: `degree(y**2+x**3, 1)` -> `ValueError: For multivariate expressions, an explicit generator must be given for gen, not an integer.`

Commit:
degree: Tests for errors with numeric gens

Tests that errors are raised for numeric non-int gens and int gens with multivariate polys

Commit:
degree: added strict=True to calls to sympify

user234683 commented:
> Can you see any ways to improve the following?

I was able to implement the suggested changes with a few tweaks to the code you suggested earlier.

user234683 commented:
> In addition, I think it's not a good idea to let a multivariate expression get an index to a generator.

A lot of code seems to rely on using `degree(expr)` (which implicity uses index 0, the first variable alphabetically), hence the failed tests.

Commit:
degree: reverted tests for multivariate w/ numeric gens

Commit:
degree: re-added numeric gens for multivariates

smichr committed:
tweak: only recast if original f was Poly

If the original f was not a Poly then there is not need for
recasting since p is already made without explicit generators.

smichr commented:
Looks good. Rather than describe the last tweak, I just added it. See commit message.

smichr commented:
I think that any code that relies on `gen=0` as the default for a multivariate polynomial should pass the generator but that can be a separate issue.

smichr committed:
Update polytools.py

I think only the first change is necessary since a numeric f will already have been handled
and anything that couldn't be sympified strictly would have raised an error. But...to be
safe the isisntance is being used.

smichr commented:
ping @jksuom -- the blocking PR has been merged so I think this is ready to finish off when you get the chance.

jksuom commented on sympy/polys/polytools.py:
```diff
@@ -4408,7 +4408,7 @@ def degree(f, gen=0):
     """
 
-    if f.is_Poly:
+    if isinstance(f, Poly):
```
This change is probably not necessary. I think that the `.is_Poly` attribute was introduced for exactly the same purpose. (See `git show 39f4887c2`.)

user234683 replied:
Needs to be `sympify`'d first to use `.is_Poly` (otherwise it fails for ints)

smichr committed:
sympify f and use is_Poly

smichr committed:
sympyify -> sympify

sigh. Note to self: don't use the online editor to make changes to PRs
